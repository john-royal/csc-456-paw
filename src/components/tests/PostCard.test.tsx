import PostCard from "../PostCard.tsx";
import renderer from "react-test-renderer";
import '@testing-library/jest-dom'
import {fireEvent ,render, waitFor, screen} from '@testing-library/react'
import userEvent from '@testing-library/user-event'

//types
interface Post {
    id: number;
    username: string;
    petProfilePhoto: string;
    petImage: string;
    likeCount: number;
    commentCount: number;
}
  
interface PostCardProps {
    post: Post;
}
  
const aPost: Post = {
    id: 123,
    username: "Borat",
    petProfilePhoto: "cat.jpg",
    petImage: "pfp.png",
    likeCount: 100,
    commentCount: 14,
}

const sendCard: PostCardProps = {
    post: aPost
}


describe('Unit Tests for PostCard component', () => {
    test('check if test matches snapshot', () => {
        //make sure that there are no cached snapshots in the __snapshots__ dir
        const component = renderer.create(<PostCard post={sendCard.post}/>);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
    
    test('check if like value is rendered', () => {
        const { getByText } = render(<PostCard post={sendCard.post}/>);
        const likeCountElem = getByText(String(aPost.likeCount));
        expect(likeCountElem).toBeInTheDocument();
    });
    
    test('check if like value is incremented', async () => {
        const { getByTestId, getByText } = render(<PostCard post={sendCard.post}/>);
        const likeCountBttn = getByTestId('like-bttn');
    
        fireEvent.click(likeCountBttn); //update the value
        const likeCount = getByText(String(aPost.likeCount + 1)); // likeCount value should be updated by now
    
        await waitFor(() => {
            expect(likeCount).toBeInTheDocument();
        }, {timeout: 5000});
    });
    
    test('check if like value is decremented', async () => {
        const { getByTestId, getByText } = render(<PostCard post={sendCard.post}/>);
        const likeCountBttn = getByTestId('like-bttn');
    
        fireEvent.click(likeCountBttn); //add a like
        fireEvent.click(likeCountBttn); //take back the like
        const likeCount = getByText(aPost.likeCount); // likeCount value should be updated by now
    
        await waitFor(() => {
            expect(likeCount).toBeInTheDocument();
        }, {timeout: 3000});
    });
    
    test('comment section renders', async () => {
        const { getByTestId } = render(<PostCard post={sendCard.post}/>);
        const createCommentbutton = getByTestId('cmnt-bttn');
    
        fireEvent.click(createCommentbutton);//open comment section...
        const postButton = getByTestId('cmnt-post-bttn'); // likeCount value should be updated by now...
    
        await waitFor(() => {
            expect(postButton).toBeInTheDocument();
        }, {timeout: 3000});
    });
    
    test('check error is thrown when recieving incomplete post.', async () => { //FIXME, this console.error's the thrown error 
        const err = () => render(<PostCard post={({} as any)}/>)
        expect(err).toThrow('Invalid post data: Essential fields are missing.');
    });  

    //cover comment issues
    test('typing into textbox', async () => {//verify that comments are being created
        render(<PostCard post={sendCard.post}/>);
        const createCommentbutton = screen.getByTestId('cmnt-bttn');
        fireEvent.click(createCommentbutton);//open comment section...

        const txtArea = screen.getByTestId('cmnt-text-area');//get the txtArea
        const someTxt = 'Hello World I have been written as a test :D'

        userEvent.type(txtArea, someTxt);
        await waitFor(() => {
            expect(txtArea).toHaveValue(someTxt);
        }, {timeout: 3000});
    });

    test('adding a comment', async () => {
        render(<PostCard post={sendCard.post}/>);
        const createCommentButton = screen.getByTestId('cmnt-bttn');
        fireEvent.click(createCommentButton);//open comment section...

        const txtArea = screen.getByTestId('cmnt-text-area');//get the txtArea
        const someTxt = 'Hello World I have been written as a test :D';
        const someTxtRgx = /Hello World I have been written as a test :D/i;
        await userEvent.type(txtArea, someTxt);//write in the txtArea

        const postCommentButton = screen.getByTestId('cmnt-post-bttn');
        await userEvent.click(postCommentButton);//post comment

        await waitFor(() => {
            const addedText = screen.getByText(someTxtRgx);//regex to identify the text.
            expect(addedText).toBeInTheDocument();
        }, {timeout: 3000});
    });
});


