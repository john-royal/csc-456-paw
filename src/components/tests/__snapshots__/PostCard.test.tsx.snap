// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit Tests for PostCard component check if test matches snapshot 1`] = `
<div
  className="mb-4 rounded-md border border-gray-200 p-4 bg-white"
>
  <div
    className="mb-2 flex items-center"
  >
    <img
      alt="Pet Profile"
      className="mr-2 h-12 w-12 rounded-full"
      src="cat.jpg"
    />
    <span>
      Borat
    </span>
  </div>
  <img
    alt="Pet"
    className="mb-2 h-full w-full rounded-md object-cover"
    src="pfp.png"
  />
  <div
    className="flex items-center justify-between"
  >
    <button
      className="mr-2 flex items-center rounded-md px-4 py-2 bg-gray-300 text-gray-900"
      data-testid="like-bttn"
      onClick={[Function]}
    >
      <svg
        className="mr-1"
        data-testid="like-count"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
        />
      </svg>
       
      100
    </button>
    <button
      className="flex items-center rounded-md bg-gray-300 px-4 py-2"
      data-testid="cmnt-bttn"
      onClick={[Function]}
    >
      <svg
        className="mr-1"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 32C114.6 32 0 125.1 0 240c0 49.6 21.4 95 57 130.7C44.5 421.1 2.7 466 2.2 466.5c-2.2 2.3-2.8 5.7-1.5 8.7S4.8 480 8 480c66.3 0 116-31.8 140.6-51.4 32.7 12.3 69 19.4 107.4 19.4 141.4 0 256-93.1 256-208S397.4 32 256 32z"
        />
      </svg>
       
      14
    </button>
  </div>
</div>
`;
